### PR TITLE
Add switching logic for rest api handlers for list_artifacts

### LINKED ISSUE #5377: [BUG] Server list artifacts http 500 error for proxied uploads
Thank you for submitting an issue. Please refer to our [issue policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md) for additional information about bug reports. For help with debugging your code, please refer to [Stack Overflow](https://stackoverflow.com/questions/tagged/mlflow).

**Please fill in this bug report template to ensure a timely and thorough response.**

### Willingness to contribute
The MLflow Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the MLflow code base?

- [ ] Yes. I can contribute a fix for this bug independently.
- [x] Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.
- [ ] No. I cannot contribute a bug fix at this time.

### System information
- **Have I written custom code (as opposed to using a stock example script provided in MLflow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: `python:3.9.9-slim-bullseye` docker image
- **MLflow installed from (source or binary)**: binary
- **MLflow version (run ``mlflow --version``)**: 1.23.0
- **Python version**: 3.9.9
- **npm version, if running the dev UI**:
- **Exact command to reproduce**: See below

### Describe the problem

I have a mlflow server setup which is working fine except from listing artifacts from the ui after having uploaded them.

### Code to reproduce issue

I'm running mlflow server as follows:

```bash
mlflow  server \
      --host 0.0.0.0 \
      --port 5000 \
      --workers 4 \
      --backend-store-uri postgresql://$DATABASE_USER:$DATABASE_PASSWORD@postgres:5432/mlflow?sslmode=disable \
      --artifacts-destination s3://my-bucket/artifacts \
      --serve-artifacts \
      --gunicorn-opts "--access-logfile -"
```

And my test experiment is:

```python
import random
import mlflow

mlflow.set_tracking_uri('http://localhost:5000')
mlflow.set_experiment('test-1')

def run_experiment():
    mlflow.log_param('name', 'Peter')
    mlflow.log_param('age', 20 + random.random() * 30)

    output_path = 'output.txt'
    with open(output_path, 'w') as f:
        f.write('Something {}'.format(random.randint(30, 50)))

    mlflow.log_artifact(output_path)


with mlflow.start_run():
    run_experiment()
```

### Other info / logs

Error screenshot:
(I'm getting error 500 on http://localhost:5000/ajax-api/2.0/preview/mlflow/artifacts/list?run_uuid=a86a0e638a1748d3a48587f2d1e00488)

Screen Shot 2022-02-14 at 13 42 56

Server log:

```
2022/02/14 16:44:25 ERROR mlflow.server: Exception on /ajax-api/2.0/preview/mlflow/artifacts/list [GET]
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 2073, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1518, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1516, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1502, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/usr/local/lib/python3.9/site-packages/mlflow/server/handlers.py", line 238, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/mlflow/server/handlers.py", line 298, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/mlflow/server/handlers.py", line 571, in _list_artifacts
    artifact_entities = _get_artifact_repo(run).list_artifacts(path)
AttributeError: 'Response' object has no attribute 'list_artifacts'
172.22.0.1 - - [14/Feb/2022:16:44:25 +0000] "GET /ajax-api/2.0/preview/mlflow/artifacts/list?run_uuid=a86a0e638a1748d3a48587f2d1e00488 HTTP/1.1" 500 290 "http://localhost:5000/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/97.0.4692.99 Safari/537.36"
```

### Fix idea

In the `mlflow/server/handlers.py` file, in the `_list_artifacts()` method, I've tried using `_get_artifact_repo_mlflow_artifacts()` to fetch the repository instead of trying to fetch it from the run with `_get_artifact_repo(run)` and seems to have worked. Note that the run artifact's schema is `mlflow-artifacts:/` which, to my point of view, does not make sense to be used from the server side.

### What component(s), interfaces, languages, and integrations does this bug affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [x] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [x] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by BenWilson2**
Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

## What changes are proposed in this pull request?

Fix REST API handler to support proxied artifact mode of mlflow server (server started with ``--serve-artifacts``) so that a REST get on ``artifacts/list`` returns the correct values from the ``_list_artifacts_mlflow_artifacts()`` handler instead of the standard ``_list_artifacts()`` handler. 

## How is this patch tested?

Unit tests

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/tracking/test_mlflow_artifacts.py [hunk 1]
```diff
@@ -245,3 +246,49 @@ def test_mlflow_artifacts_example(tmpdir):
...
     )
+
+
+def get_artifacts_from_rest_api(url, run_id, path=None):
+    if path:
+        resp = requests.get(url, params={"run_id": run_id, "path": path})
+    else:
+        resp = requests.get(url, params={"run_id": run_id})
+    assert resp.status_code == 200
+    return json.loads(resp.content.decode("utf-8"))
```
Comment (harupy) [inline tests/tracking/test_mlflow_artifacts.py:257]:
```suggestion
    resp.raise_for_status()
    return resp.json()
```

nit: can we use `.raise_for_status` and `.json`?

**Comment by harupy** [comment]
https://github.com/mlflow/mlflow/blob/2078937040654cf79d7c4c065f994ba537568561/mlflow/server/handlers.py#L610-L613

Should we remove `_disable_if_artifacts_only` applied to `_get_artifact_repo` because it should be applied to endpoint handlers such as `_list_artifacts`?

File: mlflow/server/handlers.py [hunk 2]
```diff
@@ -144,6 +145,20 @@ def _get_artifact_repo_mlflow_artifacts():
     return _artifact_repo
 
 
+def _get_serve_artifacts_mode():
+    """
+    Set whether the MLflow server is operating in ``--serve-artifacts`` mode for artifact listing
+    endpoint resolution.
```
Comment (harupy) [inline mlflow/server/handlers.py:151]:
```suggestion
    Initializes `_serve_artifacts_mode` that indicates whether the MLflow server is operating in ``--serve-artifacts`` mode for artifact listing, and returns it.
    endpoint resolution.
```

Nit: can we fix the docstring to indicate this function returns something?

**Review by harupy — APPROVED** [review]
Left a few minor comments, otherwise LGTM

File: tests/tracking/test_mlflow_artifacts.py (hunk 1 above)
Comment (BenWilson2) [reply tests/tracking/test_mlflow_artifacts.py:257]: definitely!

File: mlflow/server/handlers.py (hunk 2 above)
Comment (BenWilson2) [reply mlflow/server/handlers.py:151]: changed

**Comment by BenWilson2** [comment]
> https://github.com/mlflow/mlflow/blob/2078937040654cf79d7c4c065f994ba537568561/mlflow/server/handlers.py#L610-L613
> 
> Should we remove `_disable_if_artifacts_only` applied to `_get_artifact_repo` because it should be applied to endpoint handlers such as `_list_artifacts`?
> 
> (Found this while reading the stack trace attached in #5377)

Great catch. We definitely should not be disabling this endpoint in ``-serve-artifacts`` mode.

File: mlflow/server/handlers.py [hunk 3]
```diff
@@ -560,6 +575,16 @@ def _search_runs():
 @catch_mlflow_exception
 @_disable_if_artifacts_only
 def _list_artifacts():
+
+    if _get_serve_artifacts_mode():
+        return _list_artifacts_mlflow_artifacts()
```
Comment (dbczumar) [inline mlflow/server/handlers.py:580]: I tested this out manually and discovered that this approach isn't quite correct. `ListArtifactsMlflowArtifacts` interprets the `path` parameter differently than `ListAritfacts`:

- `ListArtifactsMlflowArtifacts` interprets the `path` relative to the root directory of the configured `--artifacts-destination`. For example, supplying `"path": "1"` should list artifacts under `mlflow-artifacts:/1`.

- `ListArtifacts` interprets the `path` relative to the root artifact storage location of the run corresponding to the specified `run_id`, which is a *subdirectory* of the `--artifacts-destination`. Example: For a run with artifact root location `mlflow-artifacts:/0/60e69bf052914090b92f8b7bf1ceb0b9/artifacts`, a `ListArtifacts` request with `"path": "1"` should list artifacts under `mlflow-artifacts:/0/60e69bf052914090b92f8b7bf1ceb0b9/artifacts/1`.

This currently results in recursive listing behavior in the UI where every `ListArtifacts` call returns a list of artifacts from the root of `mlflow-artifacts:/`:

https://user-images.githubusercontent.com/39497902/155090740-2c2ba882-ea1c-4655-b92c-48207b60fa4d.mov


To fix this, we need to fetch the artifact root location of the specified run and forward it to `_list_artifacts_mlflow_artifacts()`.

**Review by dbczumar — CHANGES_REQUESTED** [review]
Looks great! Almost there! Manual testing uncovered https://github.com/mlflow/mlflow/pull/5393/files#r811678201

File: mlflow/server/handlers.py (hunk 3 above)
Comment (BenWilson2) [reply mlflow/server/handlers.py:588]: Fix is in. Validated (to be extra certain) with manual docker env checks with a remote fs container, and separate artifact server / tracking server instances. Artifact listing from client and REST API access looks good and returns correct pathing for a submitted runid listing and runid + path listings.

File: mlflow/server/handlers.py [hunk 4]
```diff
@@ -560,14 +599,61 @@ def _search_runs():
...
+        return _list_artifacts_proxy(run, path)
+    else:
+        return _list_artifacts_non_proxy(run, path)
+
+
+def _append_path_to_artifact_uri(artifact_uri, path):
+
+    parsed_uri = urllib.parse.urlparse(artifact_uri)
+    if parsed_uri.scheme.startswith("http"):
+        _, _, run_path = parsed_uri.path.split("artifacts", 2)
```
Comment (dbczumar) [inline mlflow/server/handlers.py:623]: What happens if the server admin operates the MLflow server using some prefix that contains the word `artifacts`? E.g. the routes become `http://myurl.com/artifacts/api/2.0/mlflow-artifacts/artifacts/...`?

File: mlflow/server/handlers.py [hunk 5]
```diff
@@ -560,14 +599,61 @@ def _search_runs():
...
+def _list_artifacts_proxy(run, path):
+
+    artifact_uri = run.info.artifact_uri
+
+    qualified_path = _append_path_to_artifact_uri(artifact_uri, path)
+
+    artifact_repo = _get_artifact_repo_mlflow_artifacts()
+    files = []
+    for file_info in artifact_repo.list_artifacts(qualified_path):
+        basename = posixpath.basename(file_info.path)
```
Comment (dbczumar) [inline mlflow/server/handlers.py:642]: ListArtifacts always returns paths relative to the run artifact root. This should be `posixpath.join(path, basename)`.

File: mlflow/server/handlers.py [hunk 6]
```diff
@@ -131,19 +136,35 @@ def _get_sqlalchemy_store(cls, store_uri):
...
+    from mlflow.server import SERVE_ARTIFACTS_ENV_VAR
+
+    global _serve_artifacts_mode
+    if _serve_artifacts_mode is None:
+        environment_flag = os.environ.get(SERVE_ARTIFACTS_ENV_VAR, "false") == "true"
+        artifact_repo = _get_artifact_repo_mlflow_artifacts()
+        _serve_artifacts_mode = environment_flag and isinstance(
+            artifact_repo,
+            (HttpArtifactRepository, MlflowArtifactsRepository, LocalArtifactRepository),
+        )
```
Comment (dbczumar) [inline mlflow/server/handlers.py:164]: Do we need to check this? When handling a ListArtifacts or `get-artifact` request, the MLflow server should only resolve a proxied artifact URI to an underlying location and use that underlying location to serve requests if the following are true:

1. The server is configured to run in proxied mode (i.e. the `SERVE_ARTIFACTS_ENV_VAR` os env var is set to `true`)

2. The specified run has a proxied artifact URI. We don't seem to be checking this here; instead, we're checking the artifact repo type associated with the configured artifact destination.

**Comment by Neethugkp** [comment]
Could you please check the below error.

HTTPError: 404 Client Error: Not Found for url: https://<api-gateway-url>/<instance-name>/api/2.0/mlflow-artifacts/artifacts/2/6aec34276944747ad9bc1218784a47h/artifacts/output.txt

----------
Mlflow version used  is 1.25.1

Artifacts are not getting stored in artifact location

**Comment by BenWilson2** [comment]
@Neethugkp could you share some repro code here with how you're calling the `log_artifacts()` API as well as the tracking server initialization configuration? 
That scheme is rather invalid: `https:////`

**Comment by BenWilson2** [comment]
Could you share the tracking server initialization configuration and the version of mlflow that the tracking server was started with?

**Comment by Neethugkp** [comment]
@BenWilson2 
Mlflow version : 1.25.1 
Implemented scenario 5 based on the document 
Initialization command :- 
CMD mlflow server --host 0.0.0.0 \
  --backend-store-uri "${BACKEND_URI}" \
  --artifacts-destination "${ARTIFACT_ROOT}/mlartifacts" \
  --serve-artifacts \
  --gunicorn-opts "--log-level debug" 

"${ARTIFACT_ROOT} is s3:\\bucketname.

Here MLflow is deployed in Openshift and if the hosted url is used as tracking uri at client side - Artifacts gets uploaded in storage, aswell as it is available in mlflow tracking server. It works . No issues faced here

The hosted url is then mapped into api gateway and api url will be set as tracking uri at client.
when code is executed it shows  404 error for missing artifacts.

Based on my analysis its due to the missing api method to upload artifacts. But its not defined in Mlflow document also
eg: PUT //api/2.0/mlflow-artifacts/artifacts/2/0571703baff649fba56057885ce55f37/artifacts/output.txt

Please let me know if the analysis is right. And how can this be solved.

**Comment by BenWilson2** [comment]
Is your privatelink set up to allow the file streaming functionality of --serve-artifacts? If you can connect directly to the tracking server and everything works with your configuration but there's an issue when you use api gateway to pass the artifact handling artifact logging, then there is likely something that needs to be adjusted with your VPC configuration with PrivateLink (or, if you're not using that, allowing CIDR range authorization and setting up IAM so that the tracking server can proxy the artifact management through the appropriate means). 
I'm afraid that's about as much help as I can give.

**Comment by Neethugkp** [comment]
@BenWilson2 ,


Issue is solved by configuring right datatype and parameter in api method configuration.
Thank you

**Pull request closed without merging**